From 072d9283aa7182199403dae01c24da84d05a8479 Mon Sep 17 00:00:00 2001
From: Heinz Wrobel <Heinz.Wrobel@freescale.com>
Date: Mon, 28 Jan 2019 16:05:10 +0200
Subject: [PATCH] kernel: Added phy ioctl support to the DPAA2 dpmac driver

This is needed to debug PHY registers while Linux is up and running.
Setting CONFIG_FSL_DPAA2_MAC_NETDEVS will make the dpmac's
available to the user and then the ioctl will permit checking on
the phy setup

Signed-off-by: Heinz Wrobel <Heinz.Wrobel@freescale.com>
Signed-off-by: Ciprian Costea <ciprianmarian.costea@nxp.com>
---
 drivers/staging/fsl-dpaa2/mac/mac.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/staging/fsl-dpaa2/mac/mac.c b/drivers/staging/fsl-dpaa2/mac/mac.c
index 3ddf344..27f80fd 100644
--- a/drivers/staging/fsl-dpaa2/mac/mac.c
+++ b/drivers/staging/fsl-dpaa2/mac/mac.c
@@ -143,6 +143,17 @@ static int dpaa2_mac_stop(struct net_device *netdev)
 	return 0;
 }
 
+static int dpaa2_mac_ioctl(struct net_device *netdev, struct ifreq *rq, int cmd)
+{
+	/*if (!netif_running(netdev))
+		return -EINVAL;*/
+
+	if (!netdev->phydev)
+		return -ENODEV;
+
+	return phy_mii_ioctl(netdev->phydev, rq, cmd);
+}
+
 #ifdef CONFIG_FSL_DPAA2_MAC_NETDEVS
 static netdev_tx_t dpaa2_mac_drop_frame(struct sk_buff *skb,
 					struct net_device *dev)
@@ -316,6 +327,7 @@ static const struct net_device_ops dpaa2_mac_ndo_ops = {
 	.ndo_stop		= &dpaa2_mac_stop,
 	.ndo_start_xmit		= &dpaa2_mac_drop_frame,
 	.ndo_get_stats64	= &dpaa2_mac_get_stats,
+	.ndo_do_ioctl		= &dpaa2_mac_ioctl,
 };
 
 static const struct ethtool_ops dpaa2_mac_ethtool_ops = {
-- 
2.7.4

